### DISCUSSION TITLE
Inline suppressions for Cppcheck are possibly harmful

### DISCUSSION THREAD

**Discussion by the-risk-taker** [nlohmann/json#4419] [General]
Cppcheck has some nasty implications when encountering "breaking" issues like `syntaxError` - resulting in not performing analysis on every translation unit with "breaking" code. Which, as a **result causes us to potentially have undetected problems** in the code. Specially if somebody do:

```shell
cppcheck ... --suppress=*:*lib/*
```

Then, when we run the analysis we think that everything is ok, because Cppcheck does not report anything, but under the hood the analysis is not performed at all.

> [!IMPORTANT]  
> Tested with Cppcheck v2.14.0 and JSON v3.11.3.

As an example project:

```text
.
├── CMakeLists.txt
├── json
│   └── json.hpp
└── main.cpp
```

With CMake:

```cmake
cmake_minimum_required(VERSION 3.21)
project(JsonAndCppcheck LANGUAGES CXX)

set(CMAKE_EXPORT_COMPILE_COMMANDS ON)

add_executable(example main.cpp json/json.hpp)
```

And program:

```c++
#include "json/json.hpp"

int main()
{
    int array[5] = {};
    array[5] = 5; // Cppcheck should report: Array 'array[5]' accessed at index 5, which is out of bounds. [arrayIndexOutOfBounds]
}
```

Calling:

```shell
cppcheck --check-level=exhaustive --project=build/compile_commands.json
```

we get:

```text
$ cppcheck --check-level=exhaustive --project=build/compile_commands.json
Checking main.cpp ...
json/json.hpp:3409:49: error: syntax error [syntaxError]
    class BinaryType = std::vector<std::uint8_t>, // cppcheck-suppress syntaxError
                                                ^
```

which is fine - we see that the issue in in the lib, but then we probably do:

```shell
cppcheck --inline-suppr --check-level=exhaustive --project=build/compile_commands.json
or
cppcheck --suppress=*:*json/* --check-level=exhaustive --project=build/compile_commands.json
```

and finally we get:

```shell
$ cppcheck  --suppress=*:*json/* --check-level=exhaustive --project=build/compile_commands.json
Checking main.cpp ...
```

thinking **that it's all fine but it's not**. 

---

I know that this is the Cppcheck issue not [json](https://github.com/nlohmann/json) code problem - **but if here, in CI, similar steps as above are done, then probably the Cppcheck analysis is broken** 🧐

After slightly breaking the code (removing the comma near the suppress comment - which is wrong, but keeps Cppcheck going) I get more expected results:

```text
$ cppcheck --check-level=exhaustive --project=build/compile_commands.json
Checking main.cpp ...
main.cpp:6:10: error: Array 'array[5]' accessed at index 5, which is out of bounds. [arrayIndexOutOfBounds]
    array[5] = 5;
         ^
json/json.hpp:21253:13: error: Found an exit path from function with non-void return type that has missing return statement [missingReturn]
            }
            ^
json/json.hpp:21276:13: error: Found an exit path from function with non-void return type that has missing return statement [missingReturn]
            }
            ^
```

---

I went to the Cppcheck to submit the issue but already somebody has done that: https://trac.cppcheck.net/ticket/12923

**Comment by danmar** [comment]
Could you try what happens when `--safety` option is used? Cppcheck shouldn't silently abort the file without any indication if there are critical errors then..

**Reply by the-risk-taker** [reply]
@danmar yes, indeed using `--safety`:

```shell
$ cppcheck --suppress=*:*json/* --check-level=exhaustive --project=build/compile_commands.json --safety
Checking main.cpp ...
json/json.hpp:3409:49: error: syntax error [syntaxError]
    class BinaryType = std::vector<std::uint8_t>, // cppcheck-suppress syntaxError
                                                ^
nofile:0:0: information: Active checkers: There was critical errors (use --checkers-report=<filename> to see details) [checkersReport]
```

I finally get the expected behaviour. Thank you!

---

Honestly, I didn't know about  `--safety` option, and I've been using Cppcheck for years. When was it added? Maybe it would be worth mentioning it in the output of  `--help`? I can do a PR.

**Reply by danmar** [reply]
The `--safety` was added ~ 1 year ago so it's a relatively new flag. Feel free to do a PR, I guess others might be interested too.. It's often not wanted when integrating cppcheck in IDEs, it cause non-zero exitcode if there is some critical error for instance.

**Reply by nlohmann** [reply]
PRs welcome!

**Reply by the-risk-taker** [reply]
Already on main: https://github.com/danmar/cppcheck/pull/7300 👈
